From 40a2da5983110fadaf2ea49304aedec9ab1a2e9b Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Tue, 20 Aug 2013 17:09:35 +0800
Subject: [PATCH 3/3] rt: arm: mmp3: a spinlock fix for irq_lock

raw spin_lock will be used in the non-preempt context, otherwise
the following calltrace will be seen as expected:

BUG: sleeping function called from invalid context at kernel/rtmutex.c:658
in_atomic(): 1, irqs_disabled(): 128, pid: 0, name: swapper/0
[<c0017820>] (unwind_backtrace+0x0/0x104) from [<c063d2cc>] (dump_stack+0x20/0x24)
[<c063d2cc>] (dump_stack+0x20/0x24) from [<c0067c8c>] (__might_sleep+0x12c/0x130)
[<c0067c8c>] (__might_sleep+0x12c/0x130) from [<c0646ba4>] (rt_spin_lock+0x2c/0x38)
[<c0646ba4>] (rt_spin_lock+0x2c/0x38) from [<c0022dcc>] (icu_mux_mask_irq+0x34/0x7c)
[<c0022dcc>] (icu_mux_mask_irq+0x34/0x7c) from [<c00b2438>] (handle_level_irq+0x108/0x134)
[<c00b2438>] (handle_level_irq+0x108/0x134) from [<c00ae9bc>] (generic_handle_irq+0x40/0x50)
[<c00ae9bc>] (generic_handle_irq+0x40/0x50) from [<c0022a88>] (hsi3_irq_demux+0x78/0x104)
[<c0022a88>] (hsi3_irq_demux+0x78/0x104) from [<c00ae9bc>] (generic_handle_irq+0x40/0x50)
[<c00ae9bc>] (generic_handle_irq+0x40/0x50) from [<c000f824>] (handle_IRQ+0x68/0xbc)
[<c000f824>] (handle_IRQ+0x68/0xbc) from [<c00084b0>] (asm_do_IRQ+0x18/0x1c)
[<c00084b0>] (asm_do_IRQ+0x18/0x1c) from [<c06475b4>] (__irq_svc+0x34/0xb4)
Exception stack(0xc08edf30 to 0xc08edf78)
df20:                                     60400100 fe282818 00000000 c090ba10
df40: c08ec000 c0953b48 c0650c0c c0910544 c090aba0 562f5842 00000000 c08edf9c
df60: c08edf68 c08edf78 c002c170 c000ff84 60000013 ffffffff
[<c06475b4>] (__irq_svc+0x34/0xb4) from [<c000ff84>] (cpu_idle+0x110/0x12c)
[<c000ff84>] (cpu_idle+0x110/0x12c) from [<c0633480>] (rest_init+0x90/0x94)
[<c0633480>] (rest_init+0x90/0x94) from [<c08498b8>] (start_kernel+0x32c/0x338)
INIT: version 2.88 booting

Signed-off-by: Zumeng.Chen <zumeng.chen@windriver.com>
---
 arch/arm/mach-mmp/irq-mmp3.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/arch/arm/mach-mmp/irq-mmp3.c b/arch/arm/mach-mmp/irq-mmp3.c
index fd1f3b3..b6e02ea 100644
--- a/arch/arm/mach-mmp/irq-mmp3.c
+++ b/arch/arm/mach-mmp/irq-mmp3.c
@@ -14,6 +14,7 @@
 #include <linux/init.h>
 #include <linux/irq.h>
 #include <linux/io.h>
+#include <linux/spinlock.h>
 
 #include <asm/irq.h>
 #include <asm/mach/irq.h>
@@ -30,7 +31,7 @@ struct icu_mux_irq_chip_data {
 	unsigned int	base;
 };
 
-static DEFINE_SPINLOCK(irq_lock);
+static DEFINE_RAW_SPINLOCK(irq_lock);
 
 static void icu_mux_mask_irq(struct irq_data *d)
 {
@@ -42,10 +43,10 @@ static void icu_mux_mask_irq(struct irq_data *d)
 		printk(KERN_ERR "Can not find chip data for mux irq %d\n", d->irq);
 		return;
 	}
-	spin_lock_irqsave(&irq_lock, flags);
+	raw_spin_lock_irqsave(&irq_lock, flags);
 	r = __raw_readl(chip_data->mask) | (1 << (d->irq - chip_data->base));
 	__raw_writel(r, chip_data->mask);
-	spin_unlock_irqrestore(&irq_lock, flags);
+	raw_spin_unlock_irqrestore(&irq_lock, flags);
 }
 
 static void icu_mux_unmask_irq(struct irq_data *d)
@@ -58,10 +59,10 @@ static void icu_mux_unmask_irq(struct irq_data *d)
 		printk(KERN_ERR "Can not find chip data for mux irq %d\n", d->irq);
 		return;
 	}
-	spin_lock_irqsave(&irq_lock, flags);
+	raw_spin_lock_irqsave(&irq_lock, flags);
 	r = __raw_readl(chip_data->mask) & ~(1 << (d->irq - chip_data->base));
 	__raw_writel(r, chip_data->mask);
-	spin_unlock_irqrestore(&irq_lock, flags);
+	raw_spin_unlock_irqrestore(&irq_lock, flags);
 }
 
 #define DEFINE_ICU_MUX_IRQ(_name_, irq_base, prefix)			\
-- 
1.7.11

